diff --git a/configure.ac b/configure.ac
index 9305eddf5d..6ec1d861e6 100644
--- a/configure.ac
+++ b/configure.ac
@@ -910,7 +910,14 @@ save_LIBS="$LIBS"
 save_LDFLAGS="$LDFLAGS"
 save_CPPFLAGS="$CPPFLAGS"
 
-LIBS="$STATIC_LIBEVENT_FLAGS $TOR_LIB_WS32 $save_LIBS"
+case "$host" in
+  *mingw*|*windows*)
+    LIBS="$STATIC_LIBEVENT_FLAGS $TOR_LIB_WS32 $TOR_LIB_IPHLPAPI $TOR_LIB_BCRYPT -ladvapi32 -lshell32 -luser32 $save_LIBS"
+    ;;
+  *)
+    LIBS="$STATIC_LIBEVENT_FLAGS $TOR_LIB_WS32 $save_LIBS"
+    ;;
+esac
 LDFLAGS="$TOR_LDFLAGS_libevent $LDFLAGS"
 CPPFLAGS="$TOR_CPPFLAGS_libevent $CPPFLAGS"
 
@@ -933,6 +940,11 @@ else
        if test "$ac_cv_search_evdns_base_new" != "none required"; then
          TOR_LIBEVENT_LIBS="$ac_cv_search_evdns_base_new $TOR_LIBEVENT_LIBS"
        fi
+       case "$host" in
+         *mingw*|*windows*)
+           TOR_LIBEVENT_LIBS="$TOR_LIBEVENT_LIBS $TOR_LIB_IPHLPAPI $TOR_LIB_BCRYPT -ladvapi32 -lshell32 -luser32 $TOR_LIB_WS32"
+           ;;
+       esac
      else
        AC_MSG_ERROR("libevent2 is required but the headers could not be found")
      fi
